### PR TITLE
Integrate Arwen 0.3.37

### PR DESCRIPTION
**Pull request by sasurobert**
In the new arwen ETHEREUM EEI was disabled.

### CODE REVIEW

File: process/smartContract/process.go [hunk 1]
```diff
@@ -305,9 +305,8 @@ func (sc *scProcessor) doExecuteSmartContractTransaction(
 		return 0, err
 	}
 
-	var consumedFee *big.Int
 	var results []data.TransactionHandler
-	results, consumedFee, err = sc.processVMOutput(vmOutput, txHash, tx, acntSnd, vmInput.CallType, vmInput.GasProvided)
+	results, err = sc.processVMOutput(vmOutput, txHash, tx, acntSnd, vmInput.CallType, vmInput.GasProvided)
```
Comment (SebastianMarian) [inline process/smartContract/process.go:309]: All the modifications from this file should not be backwards compatible and activated with epoch flag?

File: cmd/node/config/gasSchedule.toml [hunk 2]
```diff
@@ -597,3 +601,5 @@
     I64x2Load32x2U = 400
     I8x16RoundingAverageU = 200
     I16x8RoundingAverageU = 200
+    LocalAllocate = 2
+    LocalsUnmetered = 100
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply cmd/node/config/gasSchedule.toml:605]: done

File: integrationTests/vm/arwen/gasSchedule.toml [hunk 3]
```diff
@@ -590,3 +594,5 @@
     I64x2Load32x2U = 400
     I8x16RoundingAverageU = 200
     I16x8RoundingAverageU = 200
+    LocalAllocate = 2
+    LocalsUnmetered = 100
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply integrationTests/vm/arwen/gasSchedule.toml:598]: done

File: process/smartContract/builtInFunctions/claimDeveloperRewards.go [hunk 4]
```diff
@@ -51,10 +51,18 @@ func (c *claimDeveloperRewards) ProcessBuiltinFunction(
 	}
 
 	vmOutput := &vmcommon.VMOutput{GasRemaining: vmInput.GasProvided - c.gasCost}
+	outTransfer := vmcommon.OutputTransfer{
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/smartContract/builtInFunctions/claimDeveloperRewards.go:54]: all builtInFunctions are disabled right now.

File: process/smartContract/builtInFunctions/esdtTransfer.go [hunk 5]
```diff
@@ -93,12 +93,17 @@ func (e *esdtTransfer) ProcessBuiltinFunction(
 	if core.IsSmartContractAddress(vmInput.CallerAddr) {
 		// cross-shard ESDT transfer call through a smart contract - needs the storage update in order to create the smart contract result
 		esdtTransferTxData := core.BuiltInFunctionESDTTransfer + "@" + hex.EncodeToString(vmInput.Arguments[0]) + "@" + hex.EncodeToString(vmInput.Arguments[1])
-		vmOutput.OutputAccounts = make(map[string]*vmcommon.OutputAccount)
-		vmOutput.OutputAccounts[string(vmInput.RecipientAddr)] = &vmcommon.OutputAccount{
-			Address:  vmInput.RecipientAddr,
+		outTransfer := vmcommon.OutputTransfer{
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/smartContract/builtInFunctions/esdtTransfer.go:96]: all builtInFunctions are disabled right now.

File: process/smartContract/builtInFunctions/claimDeveloperRewards.go [hunk 6]
```diff
@@ -51,10 +51,18 @@ func (c *claimDeveloperRewards) ProcessBuiltinFunction(
...
+		Value:    big.NewInt(0).Set(value),
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/smartContract/builtInFunctions/claimDeveloperRewards.go:55]: This is the actual transfer. In BalanceDelta we keep the SUM of all outPut transfers

File: process/smartContract/builtInFunctions/claimDeveloperRewards.go [hunk 7]
```diff
@@ -51,10 +51,18 @@ func (c *claimDeveloperRewards) ProcessBuiltinFunction(
...
+		GasLimit: 0,
+		Data:     nil,
+		CallType: vmcommon.AsynchronousCall,
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/smartContract/builtInFunctions/claimDeveloperRewards.go:58]: yes. it was a bug. it is resolved now.

File: process/smartContract/builtInFunctions/saveUserName.go [hunk 8]
```diff
@@ -68,11 +69,15 @@ func (s *saveUserName) ProcessBuiltinFunction(
 		vmOutput := &vmcommon.VMOutput{ReturnCode: vmcommon.Ok}
 		vmOutput.OutputAccounts = make(map[string]*vmcommon.OutputAccount)
 		setUserNameTxData := core.BuiltInFunctionSetUserName + "@" + hex.EncodeToString(vmInput.Arguments[0])
-		vmOutput.OutputAccounts[string(vmInput.RecipientAddr)] = &vmcommon.OutputAccount{
-			Address:  vmInput.RecipientAddr,
+		outTransfer := vmcommon.OutputTransfer{
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/smartContract/builtInFunctions/saveUserName.go:72]: all builtInFunctions are disabled right now.

File: process/smartContract/process.go (hunk 1 above)
Comment (sasurobert) [reply process/smartContract/process.go:309]: they are. as deploy of smart contract are disabled. builtIn functions are disabled. There is no SC calling another SC right now - thus vmOutput and totalConsumedFee will be OK.

File: process/smartContract/process.go [hunk 9]
```diff
@@ -325,24 +324,104 @@ func (sc *scProcessor) doExecuteSmartContractTransaction(
...
+	}
+
+	return nil
+}
+
+func (sc *scProcessor) isSelfShard(address []byte) bool {
+	return sc.shardCoordinator.ComputeId(address) == sc.shardCoordinator.SelfId()
+}
+
+func (sc *scProcessor) computeTotalConsumedFee(
```
[the remark this answers is not in the document]
Comment (sasurobert) [reply process/smartContract/process.go]: done

File: process/smartContract/builtInFunctions/claimDeveloperRewards.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply process/smartContract/builtInFunctions/claimDeveloperRewards.go:55]: ok

File: process/smartContract/builtInFunctions/claimDeveloperRewards.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply process/smartContract/builtInFunctions/claimDeveloperRewards.go:58]: ok

File: process/smartContract/process.go (hunk 1 above)
Comment (SebastianMarian) [reply process/smartContract/process.go:309]: ok

**Review by SebastianMarian — DISMISSED** [review]

File: go.sum [hunk 10]
```diff
@@ -11,13 +11,23 @@ github.com/AndreasBriese/bbloom v0.0.0-20180913140656-343706a395b7/go.mod h1:bOv
 github.com/AndreasBriese/bbloom v0.0.0-20190306092124-e2d15f34fcf9/go.mod h1:bOvUY6CB00SOBii9/FifXqc0awNKxLFCL/+pkDPuyl8=
 github.com/BurntSushi/toml v0.3.1 h1:WXkYYl6Yr3qBf1K79EBnL4mak0OimBfB0XUf9Vl28OQ=
 github.com/BurntSushi/toml v0.3.1/go.mod h1:xHWCNGjB5oqiDr8zfno3MHue2Ht5sIBksp03qcyfWMU=
+github.com/ElrondNetwork/arwen-wasm-vm v0.3.33/go.mod h1:TktLl1iYOuio2TYGRVTUFI47QX0gEuOml7l3pyGsbog=
+github.com/ElrondNetwork/arwen-wasm-vm v0.3.37 h1:PNiO6i05TPTkm697VDfbMRQi210hwRjwhs0PX1wjXdM=
+github.com/ElrondNetwork/arwen-wasm-vm v0.3.37/go.mod h1:3NFiVRkjxleUw29pr6NLt0llekg4D29SVYr1Csj57Fc=
```
Comment (sasurobert) [inline go.sum:16]: go mod tidy

**Review by SebastianMarian — APPROVED** [review]

**Review by SebastianMarian — APPROVED** [review]

File: process/smartContract/process.go [hunk 11]
```diff
@@ -1189,20 +1273,20 @@ func (sc *scProcessor) updateSmartContractCode(
...
 	newCodeMetadata := vmcommon.CodeMetadataFromBytes(outputAccount.CodeMetadata)
 	isUpgradeable := currentCodeMetadata.Upgradeable
 	isDeployment := noExistingCode && noExistingOwner
-	isUpgrade := !isDeployment && isSenderOwner && isUpgradeable
+	isUpgrade := !isDeployment && isCodeDeployerOwner && isUpgradeable
 
 	if isDeployment {
-		stateAccount.SetOwnerAddress(transactionSender)
+		// At this point, we are under the condition "noExistingOwner"
+		stateAccount.SetOwnerAddress(outputAccount.CodeDeployerAddress)
```
Comment (andreibancioiu) [inline process/smartContract/process.go:1289]: Thinking about this again - I think we should add an extra check here, to assert that `outputAccount.CodeDeployerAddress` isn't nil (that is, to protect against eventual programming errors in Arwen).

File: process/smartContract/process.go [hunk 12]
```diff
@@ -325,24 +324,104 @@ func (sc *scProcessor) doExecuteSmartContractTransaction(
...
 
-	acntDst.AddToDeveloperReward(newDeveloperReward)
-	sc.txFeeHandler.ProcessTransactionFee(consumedFee, newDeveloperReward, txHash)
+	totalConsumedFee, totalDevRwd := sc.computeTotalConsumedFee(tx, vmOutput)
+	sc.txFeeHandler.ProcessTransactionFee(totalConsumedFee, totalDevRwd, txHash)
+
+	return vmcommon.Ok, nil
+}
+
+func (sc *scProcessor) updateDeveloperRewards(
```
[the remark this answers is not in the document]
Comment (andreibancioiu) [reply process/smartContract/process.go:339]: We need some unit tests, indeed - will we add them now or in a later PR?

**Review by andreibancioiu — CHANGES_REQUESTED** [review]
Also, please update the PR description to mention that now we have disabled ethereum ei.

File: process/smartContract/process.go (hunk 11 above)
Comment (andreibancioiu) [reply process/smartContract/process.go:1290]: Fixed.

**Review by andreibancioiu — APPROVED** [review]

**Review by SebastianMarian — APPROVED** [review]

**Pull request merged**